server/backend/device.go

```diff
@@ -0,0 +1,110 @@
+package backend
+
+import (
+	"context"
+	"github.com/TUM-Dev/Campus-Backend/model"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+	"gorm.io/gorm"
+	"gorm.io/gorm/clause"
+	"sync"
+	"time"
+)
+
+// deviceBuffer stores all recent device calls in a buffer and flushes them to the database periodically
+type deviceBuffer struct {
+	lock     sync.Mutex
+	devices  map[string]*model.Devices // key is uuid
+	interval time.Duration             // flush interval
+}
+
+func (s *CampusServer) RunDeviceFlusher() error {
+	for {
+		time.Sleep(s.deviceBuf.interval)
+		err := s.deviceBuf.flush(s.db)
+		if err != nil {
+			log.WithError(err).Error("Error flushing device buffer")
+		}
+	}
+}
+
+func (b *deviceBuffer) add(device model.Devices) {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+	if _, exists := b.devices[device.UUID]; exists {
+		b.devices[device.UUID].Counter++
+	} else {
+		b.devices[device.UUID] = &device
+	}
```

Comment:
You can simplify this:

```
device.Counter++
d, exists := b.devices[device.UUID];
if !exists {
	b.devices[device.UUID] = &device
}
```

Reply:
Then the device counter would always be 1 (we currently create a new device and insert it if it doesn't exist already, otherwise we create the device)